v5.4.1: after restart all tikv/pd/tidb, one tikv continue report error "invalid store ID 0, not found"

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->
./tikv-server -V
 TiKV
Release Version:   5.4.0
Edition:           Community
Git Commit Hash:   44d2e7e8d86566b4b270c1ae39a4956e0f74a6e8
Git Commit Branch: heads/refs/tags/v5.4.1
UTC Build Time:    2022-04-29 06:35:41
Rust Version:      rustc 1.56.0-nightly (2faabf579 2021-07-27)
Enable Features:   jemalloc mem-profiling portable sse test-engines-rocksdb cloud-aws cloud-gcp cloud-azure
Profile:           dist_release

### What operating system and CPU are you using?
<!-- If you're using Linux, you can run `cat /proc/cpuinfo` -->

### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->
tcms test plan:
[endless-range-partition-oltp-sysbench-oltp-read-write@v5.4.1]

following tidb global config has been set.

tidb_partition_prune_mode='dynamic' ,  
tidb_replica_read='follower'

### What did you expect?
tikv should works fine.

### What did happened?
Tikv heartbeat with pd disconnect after that, no region message update from pd monitor.

tikv logs:
Status { code: 2-UNKNOWN, message: \"invalid store ID 0, not found\", details: [] }))"]
tikv.log:[2022/05/10 02:28:24.557 +00:00] [ERROR] [[util.rs:460](http://util.rs:460/)] ["request failed"] [err_code=KV:PD:gRPC] [err="Grpc(RpcFailure(RpcStatus { code: 2-UNKNOWN, message: \"invalid store ID 0, not found\", details: [] }))"]
tikv.log:[2022/05/10 02:28:25.558 +00:00] [ERROR] [[util.rs:460](http://util.rs:460/)] ["request failed"] [err_code=KV:PD:gRPC] [err="Grpc(RpcFailure(RpcStatus { code: 2-UNKNOWN, message: \"invalid store ID 0, not found\", details: [] }))"]
tikv.log:[2022/05/10 02:28:25.565 +00:00] [ERROR] [[util.rs:460](http://util.rs:460/)] ["request failed"] [err_code=KV:PD:gRPC] [err="Grpc(RpcFailure(RpcStatus { code: 2-UNKNOWN, message: \"invalid store ID 0, not found\", details: [] }))"]
tikv.log:[2022/05/10 02:28:25.565 +00:00] [ERROR] [[raft_client.rs:749](http://raft_client.rs:749/)] ["resolve store address failed"] [err_code=KV:Unknown] [err="Other(\"[src/server/[resolve.rs:102](http://resolve.rs:102/)]: RpcFailure: 2-UNKNOWN invalid store ID 0, not found\")"] [msg="region_id: 134700 from_peer { id: 134703 store_id: 6 } to_peer {} region_epoch { conf_ver: 11 version: 1138 } extra_msg {}"] [store_id=0]
tikv.log:[2022/05/10 02:28:25.565 +00:00] [ERROR] [[util.rs:460](http://util.rs:460/)] ["request failed"] [err_code=KV:PD:gRPC] [err="Grpc(RpcFailure(RpcStatus { code: 2-UNKNOWN, message: \"invalid store ID 0, not found\", details: [] }))"]

--- a/components/pd_client/src/client.rs
+++ b/components/pd_client/src/client.rs
@@ -548,7 +548,7 @@ impl PdClient for RpcClient {
                     .with_label_values(&["get_region_by_id"])
                     .observe(duration_to_sec(timer.saturating_elapsed()));
                 check_resp_header(resp.get_header())?;
-                if resp.has_region() {
+                if resp.has_region() && resp.has_leader() {
                     Ok(Some((resp.take_region(), resp.take_leader())))
                 } else {
                     Ok(None)

--- a/components/raftstore/src/store/worker/pd.rs
+++ b/components/raftstore/src/store/worker/pd.rs
@@ -1609,9 +1609,11 @@ where
         let f = async move {
             match resp.await {
                 Ok(Some((region, leader))) => {
-                    let msg = CasualMessage::QueryRegionLeaderResp { region, leader };
-                    if let Err(e) = router.send(region_id, PeerMsg::CasualMessage(msg)) {
-                        error!("send region info message failed"; "region_id" => region_id, "err" => ?e);
+                    if leader.get_store_id() != 0 {
+                        let msg = CasualMessage::QueryRegionLeaderResp { region, leader };
+                        if let Err(e) = router.send(region_id, PeerMsg::CasualMessage(msg)) {
+                            error!("send region info message failed"; "region_id" => region_id, "err" => ?e);
+                        }
                     }
                 }
                 Ok(None) => {}

--- a/components/test_raftstore/src/pd.rs
+++ b/components/test_raftstore/src/pd.rs
@@ -1297,10 +1297,6 @@ impl TestPdClient {
         self.trigger_leader_info_loss.store(true, Ordering::SeqCst);
     }
 
-    pub fn stop_leader_info_loss(&self) {
-        self.trigger_leader_info_loss.store(false, Ordering::SeqCst);
-    }
-
     pub fn shutdown_store(&self, store_id: u64) {
         match self.cluster.write() {
             Ok(mut c) => {

--- a/components/test_raftstore/src/pd.rs
+++ b/components/test_raftstore/src/pd.rs
@@ -810,6 +810,7 @@ pub struct TestPdClient {
     tso: AtomicU64,
     trigger_tso_failure: AtomicBool,
     feature_gate: FeatureGate,
+    trigger_leader_info_loss: AtomicBool,
 }
 
 impl TestPdClient {
@@ -824,6 +825,7 @@ impl TestPdClient {
             is_incompatible,
             tso: AtomicU64::new(1),
             trigger_tso_failure: AtomicBool::new(false),
+            trigger_leader_info_loss: AtomicBool::new(false),
             feature_gate,
         }
     }
@@ -1291,6 +1293,14 @@ impl TestPdClient {
         self.trigger_tso_failure.store(true, Ordering::SeqCst);
     }
 
+    pub fn trigger_leader_info_loss(&self) {
+        self.trigger_leader_info_loss.store(true, Ordering::SeqCst);
+    }
+
+    pub fn stop_leader_info_loss(&self) {
+        self.trigger_leader_info_loss.store(false, Ordering::SeqCst);
+    }
+
     pub fn shutdown_store(&self, store_id: u64) {
         match self.cluster.write() {
             Ok(mut c) => {
@@ -1433,7 +1443,11 @@ impl PdClient for TestPdClient {
         let cluster = self.cluster.rl();
         match cluster.get_region_by_id(region_id) {
             Ok(resp) => {
-                let leader = cluster.leaders.get(&region_id).cloned().unwrap_or_default();
+                let leader = if self.trigger_leader_info_loss.load(Ordering::SeqCst) {
+                    new_peer(0, 0)
+                } else {
+                    cluster.leaders.get(&region_id).cloned().unwrap_or_default()
+                };
                 Box::pin(ok(resp.map(|r| (r, leader))))
             }
             Err(e) => Box::pin(err(e)),

--- a/tests/integrations/raftstore/test_replica_read.rs
+++ b/tests/integrations/raftstore/test_replica_read.rs
@@ -210,19 +210,14 @@ fn test_read_hibernated_region() {
     cluster.stop_node(2);
     cluster.run_node(2).unwrap();
 
-    let dropped_msgs = Arc::new(Mutex::new(Vec::new()));
-    let (tx, rx) = mpsc::sync_channel(1);
+    let store2_sent_msgs = Arc::new(Mutex::new(Vec::new()));
     let filter = Box::new(
-        RegionPacketFilter::new(1, 3)
-            .direction(Direction::Recv)
-            .reserve_dropped(Arc::clone(&dropped_msgs))
-            .set_msg_callback(Arc::new(move |msg: &RaftMessage| {
-                if msg.has_extra_msg() {
-                    tx.send(msg.clone()).unwrap();
-                }
-            })),
+        RegionPacketFilter::new(1, 2)
+            .direction(Direction::Send)
+            .reserve_dropped(Arc::clone(&store2_sent_msgs)),
     );
-    cluster.sim.wl().add_recv_filter(3, filter);
+    cluster.sim.wl().add_send_filter(2, filter);
+    cluster.pd_client.trigger_leader_info_loss();
     // This request will fail because no valid leader.
     let resp1_ch = async_read_on_peer(&mut cluster, p2.clone(), region.clone(), b"k1", true, true);
     let resp1 = resp1_ch.recv_timeout(Duration::from_secs(5)).unwrap();
@@ -231,11 +226,20 @@ fn test_read_hibernated_region() {
         "{:?}",
         resp1.get_header()
     );
-    // Wait util receiving wake up message.
-    let wake_up_msg = rx.recv_timeout(Duration::from_secs(5)).unwrap();
-    cluster.sim.wl().clear_recv_filters(3);
-    let router = cluster.sim.wl().get_router(3).unwrap();
-    router.send_raft_message(wake_up_msg).unwrap();
+    thread::sleep(Duration::from_millis(300));
+    cluster.sim.wl().clear_send_filters(2);
+    let mut has_extra_message = false;
+    for msg in std::mem::take(&mut *store2_sent_msgs.lock().unwrap()) {
+        let to_store = msg.get_to_peer().get_store_id();
+        assert_ne!(to_store, 0, "{:?}", msg);
+        if to_store == 3 && msg.has_extra_msg() {
+            has_extra_message = true;
+        }
+        let router = cluster.sim.wl().get_router(to_store).unwrap();
+        router.send_raft_message(msg).unwrap();
+    }
+    // Had a wakeup message from 2 to 3.
+    assert!(has_extra_message);
     // Wait for the leader is woken up.
     thread::sleep(Duration::from_millis(500));
     let resp2_ch = async_read_on_peer(&mut cluster, p2, region, b"k1", true, true);

